### PR TITLE
MR-23 Hero Text Overlay Addition

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Added fields to the hero image for text overlay
- New display mode to use the hero with text overlay pattern
- field values mapped to variants

# Needed By (Date)
- 7/17

# Urgency
- Medium

# Steps to Test
1. Checkout this branch
1. sync any site or install a fresh site
1. create a basic page
1. in the hero field and in the components field add a hero image
1. Dont populate any of the overlay text
1. save and view the page. validate the image displays with no color overlay and no text overlay
1. edit the page and values to the overlay title, body and/or link (any field or a combination of them)
1. save the page and validate the image now has a gradient overlay color with the text.
1. edit the page and choose a new overlay color
1. save the page and validate the color gradient changed.

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: config/default/core.entity_view_display.block_content.hs_basic_block.default.yml [hunk 1]
```diff
@@ -24,3 +24,4 @@ content:
     region: content
 hidden:
   field_hs_block_hide_title: true
+  search_api_excerpt: true
```
Comment (sherakama) [inline config/default/core.entity_view_display.block_content.hs_basic_block.default.yml:27]: Man, I hate that contrib modules do this. 

This does not seem related to this PR. Are you sure you want to have this here?

Comment (pookmish) [reply config/default/core.entity_view_display.block_content.hs_basic_block.default.yml:27]: yeah kinda annoying. it doesnt matter much. the search_api is only indexing nodes anyways. 
this is one of those things that i wanna try to figure out a method to update configs after a module is updated.

File: config/default/core.entity_view_display.paragraph.hs_hero_image.hero_with_overlay.yml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      library: null
+      disable_css: false
+      entity_classes: all_classes
+      settings:
+        pattern:
+          field_templates: default
+          variant: _field
+          variant_field: field_hs_hero_overlay_color
+          variant_field_values:
+            default: 00505c
```
Comment (sherakama) [inline config/default/core.entity_view_display.paragraph.hs_hero_image.hero_with_overlay.yml:37]: Why are some with quotes and others not?

Comment (pookmish) [reply config/default/core.entity_view_display.paragraph.hs_hero_image.hero_with_overlay.yml:37]: :man_shrugging: i wondered the same. thats just how it exported

File: config/default/core.entity_view_mode.paragraph.hero_with_overlay.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+uuid: 90e29480-6f84-4ed0-a1ff-d27964d226ca
```
Comment (sherakama) [inline config/default/core.entity_view_mode.paragraph.hero_with_overlay.yml:1]: Another view mode for all of the paragraphs that only one will use?

We had this problem with the stanford_paragraphs module and went with a generic naming convention so that each P-Type could use each view mode for something and have it be slightly more meaningful.

Comment (pookmish) [reply config/default/core.entity_view_mode.paragraph.hero_with_overlay.yml:1]: yeah i was on the fence. :person_fencing:  the nice thing is that a new entity_view_display config won't be created unless that view mode is enabled for another p-type.

File: docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml [hunk 4]
```diff
@@ -15,30 +15,15 @@ image_style_mappings:
     multiplier: 1x
     image_mapping_type: image_style
     image_mapping: hero_large
-  -
-    breakpoint_id: stanford_media.lg
```
Comment (sherakama) [inline docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml:19]: Don't need the high-density styles?

File: docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_with_text.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+dependencies:
+  config:
+    - image.style.hero_with_text_large
+    - image.style.responsive_medium
+    - image.style.responsive_small
+  module:
+    - stanford_media
+id: hero_with_text
+label: 'Hero with Text'
+image_style_mappings:
```
Comment (sherakama) [inline docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_with_text.yml:12]: Did you forget the 2x pixel density screens?

File: docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml (hunk 4 above)
Comment (pookmish) [reply docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml:19]: we had some issues with the 2x pixel density due to the size of image they provided

File: docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_with_text.yml (hunk 5 above)
Comment (pookmish) [reply docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_with_text.yml:12]: same as above :point_up:

File: docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml (hunk 4 above)
Comment (sherakama) [reply docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml:19]: Not enough pixels? 

Another approach would be to require pixel dimensions on the fields that use this hero style. Seems like you are sacrificing your end-users with Retina displays over the content author's ability to find hi-res images.

Comment (sherakama) [reply docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml:19]: Probably want to clean up the derived images and any saved config that used these if you choose to go this way.

File: docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/hs_hero_image.module [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+function hs_hero_image_entity_view_mode_alter(&$view_mode, EntityInterface $entity, $context) {
+  if ($entity->getEntityTypeId() == 'paragraph' && $entity->bundle() == 'hs_hero_image') {
+    // If the paragraph has any of the text fields populated, switch to the mode
+    // that uses the overlay pattern.
+    if (
+      $entity->get('field_hs_hero_title')->count() ||
+      $entity->get('field_hs_hero_body')->count() ||
+      $entity->get('field_hs_hero_link')->count()
+    ) {
+      $view_mode = 'hero_with_overlay';
```
Comment (sherakama) [inline docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/hs_hero_image.module:22]: Automagic vs user choice?

File: docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml (hunk 4 above)
Comment (pookmish) [reply docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/config/install/responsive_image.styles.hero_image.yml:19]: we'd to to implement some way to restrict what images are allowed to be uploaded/selected in the media library to do that. might be a difficult task but something we could do in another project.

This particular file is a copy from the `config/default` file. so existing sites dont need any cleanup or changes.

File: docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/hs_hero_image.module (hunk 6 above)
Comment (pookmish) [reply docroot/modules/humsci/hs_paragraph_types/modules/hs_hero_image/hs_hero_image.module:22]: automagic. we didnt wanna change how existing hero images looked.

**Review by sherakama — APPROVED** [review]
Code looks fine through review. I have not done any testing to see if it runs.

**Comment by boznik** [comment]
i'm in the process of empirically testing these pr's

**Review by boznik — APPROVED** [review]
This is working as expected
